File: avro/avro-annotation-processor/src/test/java/ru/tinkoff/kora/avro/annotation/processor/AvroBinaryTests.java
package ru.tinkoff.kora.avro.annotation.processor;

import org.apache.avro.generic.IndexedRecord;
import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.kora.app.annotation.processor.KoraAppProcessor;

import java.io.IOException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class AvroBinaryTests extends AbstractAvroAnnotationProcessorTest {

    @Test
    public void testReaderFromExtension() {
        compile(List.of(new KoraAppProcessor()),
            getAvroClass(),
            """
                @KoraApp
                public interface TestApp {
                  @Root
                  default String root(AvroReader<TestAvro> r) {return "";}
                }
                """);

        compileResult.assertSuccess();
        var reader = readerBinary("TestAvro");
        assertThat(reader).isNotNull();

        var testAvro = getTestAvroGenerated();
        byte[] bytes = getTestAvroAsBytes();
        IndexedRecord read = reader.readUnchecked(bytes);
        assertThatTestAvroValid(testAvro, read);
    }

    @Test
    public void testReaderBinaryFromExtension() {
        compile(List.of(new KoraAppProcessor()),
            getAvroClass(),
            """
                @KoraApp
                public interface TestApp {
                  @Root
                  default String root(@AvroBinary AvroReader<TestAvro> r) {return "";}
                }
                """);

        compileResult.assertSuccess();
        var reader = readerBinary("TestAvro");
        assertThat(reader).isNotNull();

        var testAvro = getTestAvroGenerated();
        byte[] bytes = getTestAvroAsBytes();
        IndexedRecord read = reader.readUnchecked(bytes);
        assertThatTestAvroValid(testAvro, read);
    }

    @Test
    public void testWriterFromExtension() throws IOException {
        compile(List.of(new KoraAppProcessor()),
            getAvroClass(),
            """
                @KoraApp
                public interface TestApp {
                  @Root
                  default String root(AvroWriter<TestAvro> r) {return "";}
                }
                """);

        compileResult.assertSuccess();
        var writer = writerBinary("TestAvro");
        assertThat(writer).isNotNull();

        IndexedRecord testAvro = getTestAvroGenerated();
        byte[] bytes = writer.writeBytesUnchecked(testAvro);
        IndexedRecord restored = readAsBinary(bytes);
        assertThatTestAvroValid(testAvro, restored);
    }

    @Test
    public void testWriterBinaryFromExtension() {
        compile(List.of(new KoraAppProcessor()),
            getAvroClass(),
            """
                @KoraApp
                public interface TestApp {
                  @Root
                  default String root(@AvroBinary AvroWriter<TestAvro> r) {return "";}
                }
                """);

        compileResult.assertSuccess();
        var writer = writerBinary("TestAvro");
        assertThat(writer).isNotNull();

        IndexedRecord testAvro = getTestAvroGenerated();
        byte[] bytes = writer.writeBytesUnchecked(testAvro);
        IndexedRecord restored = readAsBinary(bytes);
        assertThatTestAvroValid(testAvro, restored);
    }
}

File: avro/avro-symbol-processor/src/test/kotlin/ru/tinkoff/kora/avro/symbol/processor/AvroBinaryTests.kt
package ru.tinkoff.kora.avro.symbol.processor

import org.apache.avro.generic.IndexedRecord
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.avro.common.AvroReader
import ru.tinkoff.kora.avro.common.AvroWriter
import tinkoff.kora.binary.TestAvroBinary
import java.time.Instant

class AvroBinaryTests : AbstractAvroSymbolProcessorTest() {

    @Test
    fun testReaderAndWriterFromExtension() {
        compile0(
            """
                @KoraApp
                interface TestApp {
                    @Root
                    fun root(@AvroBinary r: AvroReader<TestAvroBinary>, @AvroBinary w: AvroWriter<TestAvroBinary>) = ""
                }
                
                """.trimIndent()
        )

        compileResult.assertSuccess()
        val reader: AvroReader<IndexedRecord> = compileResult.classLoader.readerBinary("tinkoff.kora.binary", "TestAvroBinary")
        val writer: AvroWriter<IndexedRecord> = compileResult.classLoader.writerBinary("tinkoff.kora.binary", "TestAvroBinary")
        assertThat(reader).isNotNull()
        assertThat(writer).isNotNull()

        val testAvro: IndexedRecord = TestAvroBinary.newBuilder()
            .setCluster("cluster")
            .setDate(Instant.EPOCH)
            .setDescription("descr")
            .setCounter(12345L)
            .setFlag(true)
            .build()
        val write = writer.writeBytes(testAvro)
        val read = reader.readUnchecked(write)
        assertThatTestAvroValid(testAvro, read)
    }
}
